metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py

```diff
@@ -0,0 +1,660 @@
+    BASE_URL: str = "https://api.powerbi.com/v1.0/myorg/groups"
+    ADMIN_BASE_URL: str = "https://api.powerbi.com/v1.0/myorg/admin"
+    AUTHORITY: str = "https://login.microsoftonline.com/"
+    TOP: int = 1000
+
+    def __init__(
+        self,
+        client_id: str,
+        client_secret: str,
+        tenant_id: str,
+    ):
+        self.__access_token: str = ""
+        self.__tenant_id = tenant_id
+        # Power-Bi Auth (Service Principal Auth)
+        self.__msal_client = msal.ConfidentialClientApplication(
+            client_id,
+            client_credential=client_secret,
+            authority=DataResolverBase.AUTHORITY + tenant_id,
+        )
+
+        # Test connection by generating access token
+        logger.info("Trying to connect to {}".format(self._get_authority_url()))
+        self.get_access_token()
+        logger.info("Connected to {}".format(self._get_authority_url()))
+        self._request_session = requests.Session()
+        # set re-try parameter for request_session
+        self._request_session.mount(
+            "https://",
+            HTTPAdapter(
+                max_retries=Retry(
+                    total=3,
+                    backoff_factor=1,
+                    allowed_methods=None,
+                    status_forcelist=[429, 500, 502, 503, 504],
+                )
+            ),
+        )
+
+    @abstractmethod
+    def get_groups_endpoint(self) -> str:
+        pass
+
+    @abstractmethod
+    def get_dashboards_endpoint(self, workspace: Workspace) -> str:
+        pass
+
+    @abstractmethod
+    def get_reports_endpoint(self, workspace: Workspace) -> str:
+        pass
+
+    @abstractmethod
+    def get_tiles_endpoint(self, workspace: Workspace, dashboard_id: str) -> str:
+        pass
+
+    @abstractmethod
+    def _get_pages_by_report(self, workspace: Workspace, report_id: str) -> List[Page]:
+        pass
+
+    @abstractmethod
+    def get_dataset(
+        self, workspace_id: str, dataset_id: str
+    ) -> Optional[PowerBIDataset]:
+        pass
+
+    @abstractmethod
+    def get_users(self, workspace_id: str, entity: str, entity_id: str) -> List[User]:
+        pass
+
+    def _get_authority_url(self):
+        return "{}{}".format(DataResolverBase.AUTHORITY, self.__tenant_id)
+
+    def get_access_token(self):
+        if self.__access_token != "":
+            logger.debug("Returning the cached access token")
+            return self.__access_token
+
+        logger.info("Generating PowerBi access token")
+
+        auth_response = self.__msal_client.acquire_token_for_client(
+            scopes=[DataResolverBase.SCOPE]
+        )
+
+        if not auth_response.get("access_token"):
+            logger.warning(
+                "Failed to generate the PowerBi access token. Please check input configuration"
+            )
+            raise ConfigurationError(
+                "Powerbi authorization failed . Please check your input configuration."
```

Comment:
"Failed to retrieve access token for PowerBI principal. Please verify your configuration values".

Reply:
done